scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala

```diff
@@ -0,0 +1,130 @@
+      s"\nSource:\n${sourceType}\nTarget:\n${targetType}\nResult:\n${groupResult}")
+    new MessageType(groupResult.getName, groupResult.getFields)
+  }
+
+  /**
+   * Traverse source/target schemas and format nodes of list or map.
+   * The formatting is not to one-to-one node swapping from source to target,
+   * this is because the subset fields of source node and its optional/required must
+   * be maintained in the formatted result.
+   */
+  private def formatForwardCompatibleType(sourceType: Type, targetType: Type): Type = {
+    (findCollectionGroup(sourceType), findCollectionGroup(targetType)) match {
+      case (Some(sourceGroup: ListGroup), Some(targetGroup: ListGroup)) =>
+        formatForwardCompatibleCollectionGroup[ListGroup](sourceGroup, targetGroup)
+      case (Some(sourceGroup: MapGroup), Some(targetGroup: MapGroup)) =>
+        formatForwardCompatibleCollectionGroup[MapGroup](sourceGroup, targetGroup)
+      case _ if sourceType.isPrimitive || targetType.isPrimitive => // Base case
+        sourceType
+      case _ => // Field projection
+        val sourceGroup = sourceType.asGroupType
+        val targetGroup = targetType.asGroupType
+        val resultFields = sourceGroup.getFields.asScala.map { sourceField =>
+          if (!targetGroup.containsField(sourceField.getName)) {
+            if (!sourceField.isRepetition(Repetition.OPTIONAL)) {
+              throw new DecodingSchemaMismatchException(
+                s"Found non-optional source field ${sourceField.getName}:\n$sourceField\n\n" +
+                  s"not present in the given target type:\n${targetGroup}"
+              )
+            }
+            sourceField
+          }
+          else {
+            val fieldIndex = targetGroup.getFieldIndex(sourceField.getName)
+            val targetField = targetGroup.getFields.get(fieldIndex)
+            formatForwardCompatibleType(sourceField, targetField)
+          }
+        }
+        sourceGroup.withNewFields(resultFields.asJava)
+    }
+  }
+
+  private def formatForwardCompatibleCollectionGroup[T <: CollectionGroup](sourceGroup: T,
+                                                                           targetGroup: T)
+                                                                          (implicit t: ClassTag[T]): GroupType = {
+
+    val formatter = t.runtimeClass.asInstanceOf[Class[T]] match {
+      case c if c == classOf[MapGroup] => ParquetMapFormatter
+      case c if c == classOf[ListGroup] => ParquetListFormatter
+    }
+    val formattedRepeated = formatter.formatForwardCompatibleRepeatedType(
+      sourceGroup.repeatedType,
+      targetGroup.repeatedType,
+      formatForwardCompatibleType(_, _))
+    // Wrap the formatted repeated type in its original group.
+    // This maintains the field name, and optional/required information
+    sourceGroup.groupType.withNewFields(formattedRepeated)
+  }
+
+  private def findCollectionGroup(typ: Type): Option[CollectionGroup] = {
```

Comment:
Naming nit:

```suggestion
  private def extractCollectionGroup(typ: Type): Option[CollectionGroup] = {
```

"find" implies that the return values will be a subset of the input, while "extract" (which the underlying functions use) suggests that the whole input will be considered.